clients/githubrepo/client.go

```diff
@@ -72,27 +113,114 @@ func (client *Client) InitRepo(owner, repoName string) error {
 	defer resp.Body.Close()
 
 	// Create a temp file. This automaticlly appends a random number to the name.
-	repoFile, err := ioutil.TempFile("", repoFilename)
+	repoFile, err := ioutil.TempFile(client.tempDir, repoFilename)
 	if err != nil {
-		return fmt.Errorf("ioutil.TempFile: %w", err)
+		return fmt.Errorf("error during ioutil.TempFile in githubrepo: %w", err)
 	}
 	defer repoFile.Close()
 
-	client.tarball = repoFile.Name()
-
 	if _, err := io.Copy(repoFile, resp.Body); err != nil {
-		return fmt.Errorf("io.Copy: %w", err)
+		return fmt.Errorf("error during io.Copy in githubrepo tarball: %w", err)
 	}
+	client.tempTarFile = repoFile.Name()
 	return nil
 }
 
-func (client *Client) GetRepoArchiveReader() (io.ReadCloser, error) {
+// nolint: gocognit
+func (client *Client) extractTarball() error {
 	// nolint: gomnd
-	archiveReader, err := os.OpenFile(client.tarball, os.O_RDONLY, 0o644)
+	in, err := os.OpenFile(client.tempTarFile, os.O_RDONLY, 0o644)
 	if err != nil {
-		return archiveReader, fmt.Errorf("os.OpenFile: %w", err)
+		return fmt.Errorf("error opening %s: %w", client.tempTarFile, err)
 	}
-	return archiveReader, nil
+	gz, err := gzip.NewReader(in)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", client.tempTarFile, err)
+	}
+	tr := tar.NewReader(gz)
+	for {
+		header, err := tr.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("error in tarReader.Next(): %w", err)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			dirpath, err := extractAndValidateArchivePath(header.Name, client.tempDir)
+			if err != nil {
+				return fmt.Errorf("error extracting dirpath: %w", err)
+			}
+			if dirpath == "" {
+				continue
+			}
+			// nolint: gomnd
+			if err := os.Mkdir(filepath.Join(client.tempDir, dirpath), 0o755); err != nil {
+				return fmt.Errorf("error during os.Mkdir: %w", err)
+			}
+		case tar.TypeReg:
+			if header.Size <= 0 {
+				continue
+			}
+			filenamepath, err := extractAndValidateArchivePath(header.Name, client.tempDir)
+			if err != nil {
+				return fmt.Errorf("error extracting file path: %w", err)
+			}
+			if filenamepath == "" {
+				continue
+			}
+			outFile, err := os.Create(filepath.Join(client.tempDir, filenamepath))
```

Comment:
are we guaranteed that any required dirs will come before the file entries?

Reply:
Good point. Done.